src/main/scala/pipedsl/typechecker/TypeInferenceWrapper.scala

```diff
@@ -177,25 +189,67 @@ object TypeInferenceWrapper
      {
       val inputTypes = m.inputs.map(p => p.typ)
       val modTypes = m.modules.map(m => replaceNamedType(m.typ, env))
+      val mod_subs = m.modules.zip(modTypes).foldLeft(List() :Subst)((subst, types) =>
+      compose_subst(subst, unify(types._1.typ, types._2)._1))
       val modEnv = env.add(m.name, TModType(inputTypes, modTypes, m.ret, Some(m.name)))
       val inEnv = m.inputs.foldLeft[Environment[Id, Type]](modEnv)((env, p) => env.add(p.name, p.typ))
       val pipeEnv = m.modules.zip(modTypes).foldLeft[Environment[Id, Type]](inEnv)((env, m) => env.add(m._1.name, m._2))
-      val (fixed_cmd, _, subst) = checkCommand(m.body, pipeEnv.asInstanceOf[TypeEnv], List())
+      val (fixed_cmd, _, subst) = checkCommand(m.body, pipeEnv.asInstanceOf[TypeEnv], mod_subs)
       val hash = mutable.HashMap.from(subst)
       val newMod = typeMapModule(m.copy(body = fixed_cmd).copyMeta(m), fopt_func(type_subst_map_fopt(_, hash)))
-      (modEnv, newMod)
+
+      val new_input = newMod.inputs.map(p => p.typ)
+      val new_mod_tps = newMod.modules.map(m => replaceNamedType(m.typ, env))
+
+      val fin_env = modEnv.remove(m.name)
+        .add(newMod.name, TModType(new_input, new_mod_tps, newMod.ret, Some(newMod.name)))
+      (fin_env, newMod)
      }
 
     def checkFunc(f: FuncDef, env: TypeEnv): (Environment[Id, Type], FuncDef) =
      {
       val inputTypes = f.args.map(a => a.typ)
       val funType = TFun(inputTypes, f.ret)
       val funEnv = env.add(f.name, funType)
+      val templated = mutable.HashSet.from(f.templateTypes)
+      if(templated.nonEmpty && false)
```

Comment:
why && false?

Reply:
similar to a comment. now it has left us.